### DISCUSSION TITLE
Difficulté pour programmer

### DISCUSSION THREAD

**Discussion by Alex1164** [FredM67/PVRouter-3-phase#77] [Q&A]
Bonjour,

J'ai le routeur triphasé avec trois sorties, deux sur mon chauffe-eau triphasé sans neutre (pin d3 et d4) et une troisième pas encore branchée (pin d5). Il tourne avec le programme fourni sur le site mk2pvrouter.com, téléversé avec Arduino 1.8.

J'aimerais utiliser votre programme pour pouvoir utiliser la rotation entre charges (seulement sur les pin d3 et d4).

En utilisant Arduino 2.2 sous Mac, j'ai le problème de c++17 que je n'arrive pas à résoudre.

J'ai donc installé Visual Studio Code et l'extension PlatformIO. Lorsque j'effectue la vérification du fichier cal_CTx_v_meter c'est bon mais lorsque j'effectue la vérification du fichier Mk2_3phase_RFdatalog_temp, je me retrouve avec 26 problèmes.
Pourriez-vous m'expliquer ce que je dois faire ?

<img width="960" alt="Capture d’écran 2023-10-26 à 12 10 43" src="https://github.com/FredM67/PVRouter-3-phase/assets/149060383/6232a23b-3ba4-43fa-ad2a-cf90dbefef13">

**Comment by FredM67** [comment]
Bonjour Alex,
Il ne faut pas se fier à cette analyse, très souvent erronée. Le plus simple est de compiler, si tout est ok, alors c'est ... ok !
Par ailleurs, il se peut que certaines extensions soient manquantes ....

Que dit la compilation ?
Fred

**Comment by FredM67** [comment]
En regardant plus en détails, j'ai l'impression qu'il manque des fichiers !
![image](https://github.com/FredM67/PVRouter-3-phase/assets/45318189/ca9c76a7-75e7-4752-b95f-e9b98bf20650)
Notamment le dossier "type_traits"

Pour être sûr de tout avoir, le plus simple est de clôner le dépôt en entier.

**Reply by Alex1164** [reply]
Merci beaucoup pour cette réponse rapide, c'était bien ça, il a fallu que je rajoute les fichiers manquants et que je réagence mon dossier.
J'ai téléversé le programme avec succès, par contre j'ai la première led (d3) qui clignote pour me signaler l'envoi du surplus mais j'ai également la deuxième led (d4) fixe. Cela ne me faisait pas ça avec l'ancien programme. Je ne trouve pas cela normal.

Concernant la rotation, si je ne veux faire une rotation qu'avec d3, d4 et non avec d3, d4, d5 où est-ce que je dois l'implémenter ?

**Reply by FredM67** [reply]
Alors déjà, il faut saisir les bonnes valeurs "un peu partout", et notamment les valeurs de calibration et les numéros de sortie.
Pour la rotation, il y a plusieurs modes (off, auto ou manuel).
Je n'ai pas encore prévu de config détaillée pour les rotations (liste des sorties concernées), car jusqu'à présent, ceux qui voulaient cette fonction avaient toujours des trucs sur-mesure ( genre rotation par pair, ... ).
Le code pour la rotation se trouve dans la fonction `updatePhysicalLoadStates`

**Reply by FredM67** [reply]
Si tu me donnes tous les paramètres, je peux faire les modifs...

**Comment by Alex1164** [comment]
<img width="646" alt="Capture d’écran 2023-10-26 à 16 02 32" src="https://github.com/FredM67/PVRouter-3-phase/assets/149060383/6c419772-66e8-4b73-aa5c-6eb14fafd508">

Dans calibration.h je n'ai modifié que f_powerCal (valeurs trouvées avec la procédure du site mk2pvrouter). Mais je pense que je vais refaire une calibration avec votre programme.

<img width="646" alt="Capture d’écran 2023-10-26 à 16 03 50" src="https://github.com/FredM67/PVRouter-3-phase/assets/149060383/228eaad9-84aa-422a-b5ca-f43bf65e6f4f">
<img width="646" alt="Capture d’écran 2023-10-26 à 16 04 24" src="https://github.com/FredM67/PVRouter-3-phase/assets/149060383/040e9a1a-9588-404d-b519-42d6e777a522">

Dans config.h j'ai modifié le nombre de sorties, les numéros de sorties et le mode de rotation.

A quoi sert le bouton reset (pin reset) ?

**Reply by FredM67** [reply]
Le bouton reset sert à redémarrer le routeur en cas de plantage par exemple. Pour l'instant, je n'ai jamais eu le cas. Si le mono, on peut éteindre et rallumer avec le bouton marche/arrêt, mais sur le tri, en général il est rare d'avoir la possibilité de faire un OFF-ON !

**Reply by FredM67** [reply]
A vue de nez, les paramètres me paraissent corrects.

**Comment by Alex1164** [comment]
Et en faisant une rotation avec les paires (d3, d5) et (d4, d5) est-ce que cela fonctionnerait ?

**Reply by FredM67** [reply]
C'est à dire ?
Quand je parlais de rotation par paire c'était pour quelqu'un qui avait un routeur avec 8 sorties et il voulait une rotation pour 1-2-3 et 5-6-7.
Car les sorties 1-5, 2-6 et 3-7 fonctionnaient en même temps.

**Comment by Alex1164** [comment]
Jour 1 d3 et d5 activés
Jour 2 d4 et d5 activés
Jour 3 d3 et d5 activés
etc
Je sais pas si c'est plus clair.

**Reply by FredM67** [reply]
Ah ok, mais ce n'est pas vraiment une rotation dans ce cas. Une rotation, ca serait d3-d4-d5, puis d4-d3-d5, puis d3-d4-d5, ....
Mais si tu veux un jour sur 2 désactiver en alternance d3/d4, oui, c'est faisable aussi !
Peut-être serait-il plus judicieux de faire un truc comme ceci: d3-d5-d4, d4-d5-d3, d3-d5-d4 .... non?

**Comment by Alex1164** [comment]
Oui c'est parfait d3-d5-d4, d4-d5-d3, etc
